src/ImageSharp/Formats/Gif/LzwDecoder.cs

```diff
@@ -232,6 +232,159 @@ public void DecodePixels(int minCodeSize, Buffer2D<byte> pixels)
+    {
+        // Calculate the clear code. The value of the clear code is 2 ^ minCodeSize
+        int clearCode = 1 << minCodeSize;
+
+        // It is possible to specify a larger LZW minimum code size than the palette length in bits
+        // which may leave a gap in the codes where no colors are assigned.
+        // http://www.matthewflickinger.com/lab/whatsinagif/lzw_image_data.asp#lzw_compression
+        if (minCodeSize < 2 || clearCode > MaxStackSize)
+        {
+            // Don't attempt to decode the frame indices.
+            // Theoretically we could determine a min code size from the length of the provided
+            // color palette but we won't bother since the image is most likely corrupted.
+            GifThrowHelper.ThrowInvalidImageContentException("Gif Image does not contain a valid LZW minimum code.");
+        }
+
+        int codeSize = minCodeSize + 1;
+
+        // Calculate the end code
+        int endCode = clearCode + 1;
+
+        // Calculate the available code.
+        int availableCode = clearCode + 2;
+
+        // Jillzhangs Code see: http://giflib.codeplex.com/
+        // Adapted from John Cristy's ImageMagick.
+        int code;
+        int oldCode = NullCode;
+        int codeMask = (1 << codeSize) - 1;
+        int bits = 0;
+
+        int top = 0;
+        int count = 0;
+        int bi = 0;
+        int xyz = 0;
+
+        int data = 0;
+        int first = 0;
+
+        ref int prefixRef = ref MemoryMarshal.GetReference(this.prefix.GetSpan());
+        ref int suffixRef = ref MemoryMarshal.GetReference(this.suffix.GetSpan());
+        ref int pixelStackRef = ref MemoryMarshal.GetReference(this.pixelStack.GetSpan());
+
+        for (code = 0; code < clearCode; code++)
+        {
+            Unsafe.Add(ref suffixRef, code) = (byte)code;
+        }
+
+        Span<byte> buffer = stackalloc byte[byte.MaxValue];
+        while (xyz < length)
+        {
+            if (top == 0)
+            {
+                if (bits < codeSize)
+                {
+                    // Load bytes until there are enough bits for a code.
+                    if (count == 0)
+                    {
+                        // Read a new data block.
+                        count = this.ReadBlock(buffer);
+                        if (count == 0)
+                        {
+                            break;
+                        }
+
+                        bi = 0;
+                    }
+
+                    data += buffer[bi] << bits;
+
+                    bits += 8;
+                    bi++;
+                    count--;
+                    continue;
+                }
+
+                // Get the next code
+                code = data & codeMask;
+                data >>= codeSize;
+                bits -= codeSize;
+
+                // Interpret the code
+                if (code > availableCode || code == endCode)
+                {
+                    break;
```

Comment:
This break here is the killer. A lot of properties we require for each loop follow.